File: src/189RotateArray/189RotateArray.go
package main

func reverse(nums []int, lo, hi int) {
	for lo <= hi {
		nums[lo], nums[hi] = nums[hi], nums[lo]
		lo++
		hi--
	}
}
func rotate(nums []int, k int) {
	n := len(nums)
	if n == 0 {
		return
	}
	k = ((k % n) + n) % n
	if k == 0 {
		return
	}
	reverse(nums, 0, n-1-k)
	reverse(nums, n-k, n-1)
	reverse(nums, 0, n-1)
}

func main() {

}

File: src/189RotateArray/189RotateArray.rs
impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let n = nums.len() as i32;
        if n==0 {return;}
        let k = (k % n + n) % n;
        let mut i = 0;
        let mut j = n-k-1;
        let mut l = n-1;
        while i<j {nums.swap(i as usize,j as usize);i+=1;j-=1;}
        j=n-k;
        while j<l {nums.swap(j as usize,l as usize);j+=1;l-=1;}
        i=0;
        l=n-1;
        while i<l {nums.swap(i as usize,l as usize);i+=1;l-=1;}
    }
}
